### DISCUSSION TITLE
Deployment issue

### DISCUSSION THREAD

**Discussion by mdminhaj88** [opencrvs/opencrvs-core#6558] [Technical Q&A] [answered]
### OpenCRVS Core version

1.4.0

### OpenCRVS Countryconfig version

1.4.0

### Your operating system and version

Ubuntu 22.04

### Your node version

v16.20.0

### Which of these describes your project best?

- [X] Testing OpenCRVS for business development, educational purposes or for personal interest.
- [X] A research project to evaluate, configure and test OpenCRVS as a possible solution for a government, an active brief or proposal.
- [ ] Actively installing and configuring OpenCRVS Core for production use in a country.
- [ ] Customising OpenCRVS Core code in production in a country

**Comment by mdminhaj88** [comment]
Hi @euanmillar ,

hope you are doing good, i tried to deploy the qa environment onto my server after creating the environment, running the playbook and publishing the image to the repository but in the last step of deployment of application onto the qa server ( note that i only want to deploy the qa environment without production or any other environment for now) i am given an error as seen in the screenshot

![deploy fail](https://github.com/opencrvs/opencrvs-core/assets/142885612/6e478cda-2ff2-49ba-8245-e63e46fe75dc)

while the qa server (VM) i am using is connected to the wireguard client only

![wg-2](https://github.com/opencrvs/opencrvs-core/assets/142885612/43dd4016-ba54-456a-89cf-52896f4201e2)

but now it is mandatorily asking for this environment variable, help me out to overcome this issue as i am stuck at the final step.

And if the above mentioned Env-var are mandatory then suggest me the easiest way to configure the vpn and what all changes i have to make in docker-compose files or while creating environment with the script.

I would also want you to verify that can we deploy only 'qa' environment and access the application without deploying other environments like staging and production.

**Reply by euanmillar** [reply]
Hi, thanks for raising this.  We will investigate why the deploy script is demanding these optional variables.  FYI @rikukissa ...

In the meantime, just manually create these 2 environment secrets as strings with a space. It won't break anything.  At least you will be able to deploy then.

It's definitely possible to deploy just a standalone QA or development environment.

**Reply by mdminhaj88** [reply]
hi, 
i've finally run the deploy workflow and after much processing and configuring it has finally gone under a loop of "curl: (28) Failed to connect to kibana port 5601 after 60001 ms: Timeout was reached" at the last and after a lot of attempts its failing the workflow. what may be the possible reason for this?

PFA

![deploy kibana error](https://github.com/opencrvs/opencrvs-core/assets/142885612/21c51278-44fa-4677-8196-3d051b4108c7)

Thank you

**Reply by rikukissa** [reply]
Indeed, in our default opencrvs-countryconfig template, inside the [docker-compose-qa-deploy.yml](https://github.com/opencrvs/opencrvs-countryconfig/blob/develop/infrastructure/docker-compose.qa-deploy.yml#L131), both the VPN_ADMIN_PASSWORD and SENTRY_DSN variables are defined as required variables (no default value defined). `VPN_ADMIN_PASSWORD` is there as wg-easy, the VPN server we provide as part of the QA environment needs an admin password to be used in the web UI. Like Euan said, you can set these Github secrets as empty strings.

>hi, now as i deleted the previous qa environment which i had created because i selected 'no' to the option during the script asking whether "this server is the jump server" and now i created another environment by selecting the option as yes... and maybe due to this the deploy workflow is failing at install ssh key step, whereas the provision qa environment workflow has run successfully with the same known_hosts file and settings. also when i ran the deploy workflow yesterday there was no such error of 'install-ssh-key' im not able to understand where i went wrong... can you please guide me

It seems you got past this error, correct? I notice there is one configuration error in your `qa.yml` config 
https://github.com/mdminhaj88/opencrvs-countryconfig/blob/release-v1.4.0_minhaj/infrastructure/server-setup/qa.yml#L14-L18

The `provision` user should already be existing before the provision process starts (created by you manually). This list of users is a list of human users who are maintaining the server and who the provisioning automation creates a user for. So in your case, it could look something like 

```
      - name: mdminhaj88
        ssh_keys:
          - your personal ssh public key you use for logging in        
        state: present
        sudoer: true
```

After provisioning successfully, you should verify `ssh mdminhaj88@164.52.201.146` let's you log in to the server

**Reply by mdminhaj88** [reply]
yes @rikukissa thank you for your quick response, i got past that error and i've run the deploy github action and facing the issue i have mentioned above regarding the port 5601 used by kibana.... kindly guide me through that.

And even after the github action log confirms the presence of services like opencrvs_kibana "Updating service opencrvs_kibana (id: 0venvc61sy95la9utlgkgbiu8)" still im not able to see the services on the VM when i check by running command "sudo systemctl status kibana / kibana.service / opencrvs_kibana".

Also to bring to your notice i had already installed elasticsearch, kibana, elastalert, mongodb and minio services even before provisioning the environment and all those services were accessible using the respective ports, and had  configured the credentials for all the services and had given the same credentials while running the script for environment provisioning.
later i noticed that the deploy github action is downloading and installing all these services again by itself and later on this "curl: (28) Failed to connect to kibana port 5601 after 60001 ms: Timeout was reached" error was thrown, so i thought that the services installed by myself and the services installed by the github action may be conflicting each other and then i removed all the services i had installed manually. But still the issue is persisting, i also checked the ports are not used by any service now and even after running the deploy workflow multiple times the problem still persists.

Please guide me through


thank you!

**Reply by rikukissa** [reply]
@mdminhaj88 can you give me the output of 

```
docker service ls
```

This command lists all services in Docker Swarm

and 

```
docker service ps opencrvs_kibana --no-trunc
```

```
docker service ps opencrvs_elasticsearch --no-trunc
```

These two describe the state of both kibana and elasticsearch

OpenCRVS deployment comes with elasticsearch, kibana, mongodb and so on, so you do not need to install those manually. In fact, based on what you are saying I have a hunch that it might be causing the issue you are currently facing. 
It might be that one of these dependencies fails to start if there's already an instance of the same service running on the machine reserving the same port.

**Reply by mdminhaj88** [reply]
sure @rikukissa,

please find the attached images for:
1. docker service ls

![docker service ls](https://github.com/opencrvs/opencrvs-core/assets/142885612/24c9037a-0899-48d6-94da-b4ef1e0d9640)

2. kibana and elasticsearch service

![kibana and elasticsearch service](https://github.com/opencrvs/opencrvs-core/assets/142885612/ba3a6b46-f364-4cd0-a43f-8d0aaefa480e)

thank you

**Reply by rikukissa** [reply]
Alright, all of that looks ok.. Can you add a user for me in the server. I might as well take a look. 

Please merge [this PR](https://github.com/mdminhaj88/opencrvs-countryconfig/pull/1) and provision the QA machine again.

**Reply by mdminhaj88** [reply]
Done merging the mentioned PR and added a user as well named riku and as well added you as a collaborator to my forked github repository.... @rikukissa thank you!

**Reply by rikukissa** [reply]
@mdminhaj88 can you try redeploying now? There was some kind of an issue in docker that prevented containers from connecting to each other or to the public internet. I restarted the docker service completely and networking started working

**Reply by mdminhaj88** [reply]
@rikukissa oh im surprised what caused that issue, well i tried and the deployment was successful!! but im not able to access the services through the url.

<service>.opencrvs.idencode.link for example login service

![url not working](https://github.com/opencrvs/opencrvs-core/assets/142885612/17d0e1f5-c624-42a3-a994-87caf190dea3)

**Reply by rikukissa** [reply]
@mdminhaj88 is `*.opencrvs.idencode.link` domain pointed to 164.52.201.146? At least my machine cannot resolve `login.opencrvs.idencode.link` to this IP address.

**Reply by mdminhaj88** [reply]
actually there was some issue with the A record @rikukissa .... now all the services are accessible... but the default users list mentioned in the repository, none of the user credentials are working.... its throwing incorrect username or password... can you throw some light... thank you for your constant support!

![login page](https://github.com/opencrvs/opencrvs-core/assets/142885612/fb23cb16-1cca-4fe2-b4a5-bc5c9cd239ce)

**Reply by rikukissa** [reply, marked as answer]
No worries, glad you've gotten this far. For this you have to run the "Seed data" workflow. I ran it for you [here](https://github.com/mdminhaj88/opencrvs-countryconfig/actions/runs/8114156406). Please try logging in with `k.mweene` `test` `000000` to verify it's working.

**Comment by mdminhaj88** [comment]
That was successful @rikukissa !! Thanks again
